### PR TITLE
Provide stdlib via module provider to assembler

### PR DESCRIPTION
**Pull request by vlopes11**
## Describe your changes

This initial step for the solution of #298 removes stdlibrary as dependency of assembler.

This decoupling will increase the assembler module extraction flexibility. As next step, an efficient lookup of hasher module+label will be introduced in order to fetch parsed/compiled procedures.

### CODE REVIEW

**Review by bobbinth — CHANGES_REQUESTED** [review]
Thank you! Looks good! I left a couple small comments inline.

One more general comment (could be addresses in the future): I don't love making `Assembler` generic. However, we do want to have the flexibility of defining custom logic for `ModuleProvider`.

I'm wondering if there is a way to achieve both somehow. For example, could we provide `ModuleProvider` as `dyn` reference? (I have very little experience working with such patterns).

Another potential option is making `ModuleProvider` a struct rather than a trait. This struct then could be built in a variety of ways by whoever needs to instantiate the assembler. This will make the structure less flexible we'll need to make assumptions about how `ModuleProvider` works.

Are there any other approaches?

**Comment by vlopes11** [comment]
@bobbinth there is a nice possibility to have a module provider as concrete that will have runtime loaded modules via `dyn ModuleProviderInstance`. These can be many, and this will offer further flexibility. Preparing a push with that approach

File: assembly/src/tests.rs [hunk 1]
```diff
@@ -167,45 +167,18 @@ fn program_with_proc_locals() {
...
+    let assembler = super::Assembler::default_without_module_provider();
     let source = "export.foo push.3 push.7 mul end begin push.2 push.3 add exec.foo end";
     assert!(assembler.compile(source).is_err());
 }
 
 // IMPORTS
 // ================================================================================================
 
-#[test]
-fn program_with_one_import() {
```
[the remark this answers is not in the document]
Comment (vlopes11) [reply assembly/src/tests.rs:179]: This is a test that depends on stdlib in assembly.

I will create a dummy module that will provide this code instead of requiring stdlib there

File: assembly/src/lib.rs [hunk 2]
```diff
@@ -43,6 +45,63 @@ const MODULE_PATH_DELIM: &str = "::";
...
+/// The module provider is now a simplified version of a module cache. It is expected to evolve to
+/// a general solution for the module lookup
+///
+/// TODO compute a procedure index deterministically from a module path & procedure label. The
+/// initial expected layout for the index is `[u8; 20]`. The module provider should map procedures
+/// from these indexes, and this is expected to result in code simplification + optimization for
+/// the compiler.
+pub trait ModuleProvider {
+    /// Fetch source contents provided a module path
+    fn get_source(&self, path: &str) -> Option<&str>;
```
Comment (bobbinth) [inline assembly/src/lib.rs:60]: nit: should we rename this to `get_module`? In the future, the signature here will probably change to something like:
```Rust
fn get_module(&self, index: [u8; 20]) -> Option<&ModuleAst>;
```

File: assembly/src/lib.rs [hunk 3]
```diff
@@ -4,15 +4,17 @@
 #[macro_use]
 extern crate alloc;
 
+use core::ops;
+
```
Comment (bobbinth) [inline assembly/src/lib.rs:8]: minor nit: I'd get rid of the blank line after this import.

File: assembly/src/lib.rs [hunk 4]
```diff
@@ -43,6 +45,63 @@ const MODULE_PATH_DELIM: &str = "::";
...
+    /// Fetch source contents provided a module path
+    fn get_source(&self, path: &str) -> Option<&str>;
+}
+
+#[derive(Default)]
+struct ModuleProviderController {
+    // instead of creating a map to available source, we provide a lazy load approach where the
+    // user will query for a given module path, and we will iterater the available sources
+    providers: Vec<Box<dyn ModuleProvider>>,
+}
```
Comment (bobbinth) [inline assembly/src/lib.rs:68]: What benefit do we get by adding the controller? i.e., could we define the assembler as:
```Rust
pub struct Assembler {
    module_provider: Box<dyn ModuleProvider>,
    ...
}
```
This way, it would be completely up to the `ModuleProvider` implementation to specify the logic for mapping indexes to modules. For example, if there were 2 or more libraries, a single map for all of them could be created at instantiation time, and then we won't have to iterate over all module providers.

File: assembly/src/lib.rs [hunk 5]
```diff
@@ -151,6 +222,15 @@ impl Assembler {
     // IMPORT PARSERS
     // --------------------------------------------------------------------------------------------
 
+    /// Add an instance of the provider to the internal list
+    pub fn add_module_provider<P>(&mut self, provider: P)
```
Comment (bobbinth) [inline assembly/src/lib.rs:226]: minor nit: I'd probably put this into a separate section (as it is not an "import parser"). Maybe something like "state mutators" right after "public accessors".

**Review by bobbinth — CHANGES_REQUESTED** [review]
I like this approach! Thank you! I left a few comments inline (some nits and a question about potential simplification).

File: assembly/src/lib.rs (hunk 2 above)
Comment (vlopes11) [reply assembly/src/lib.rs:60]: The original idea I had in mind is to keep it like `get_source` so we go for a minimal set of changes by reflecting the current behavior of stdlib.

The next step is to introduce the procedure index that will most likely deprecate this `get_source` as it is potentially a leaked internals of the module provider.

I'm adding a comment there explaining this approach

File: assembly/src/lib.rs (hunk 5 above)
Comment (vlopes11) [reply assembly/src/lib.rs]: Agreed, its a bit off in the parsers section. Updated!

File: assembly/src/lib.rs (hunk 3 above)
Comment (vlopes11) [reply assembly/src/lib.rs]: Sounds good! We might want to build a rustfmt.toml with the format settings so these are normalized also in the CI level.

I'll check that when I have some free cycles. Creating an issue to track it

File: assembly/src/lib.rs (hunk 4 above)
Comment (vlopes11) [reply assembly/src/lib.rs:67]: The idea is to create a very basic controller for now, but extends its functionality when the action of providing a module goes beyond fetching a source.

We will have some questions such as "should we resolve modules as `domain+name` or always full path?" that will be handled by this structure, maybe even with unit tests coverage.

We might want to delegate this problem to the implementors of `ModuleProvider` if we want to leave the procedure id resolution up to them. However, we might end with divergent module resolution that can potentially leak into the language.

wdyt?

Comment (bobbinth) [reply assembly/src/lib.rs:67]: My tendency is to do something simpler now and then add functionality to it later as we need. The model we are moving towards now is that, at the assembly level, we'll map `[u8; 20] -> ModuleAst` (maybe we should call type `[u8; 20]` procedure ID or something like that). This is primarily because procedure names will be erased at the serialized AST level, and thus we won't be able to use procedure names, full paths etc.

Basically, I'm not seeing a need to support a more generic resolution methodology than `[u8; 20] -> `ModuleAst` (the current `&str -> &str` is just a temporary step). But of course I can be mistaken :)

File: assembly/src/lib.rs [hunk 6]
```diff
@@ -104,6 +135,18 @@ impl Assembler {
...
+    // --------------------------------------------------------------------------------------------
+
+    /// Add an instance of the provider to the internal list
+    pub fn replace_module_provider<P>(&mut self, provider: P)
+    where
+        // The static lifetime will bind the type definition, and not instances of it
+        P: ModuleProvider + 'static,
+    {
+        self.module_provider = Box::new(provider);
+    }
```
Comment (bobbinth) [inline assembly/src/lib.rs:148]: We probably don't need this any more since we can set the `ModuleProvider` at construction time.

File: assembly/src/lib.rs [hunk 7]
```diff
@@ -65,28 +87,37 @@ impl Assembler {
...
         Self {
-            stdlib: StdLibrary::default(),
+            module_provider: Box::new(()),
             parsed_modules: BTreeMap::default(),
             kernel_procs: BTreeMap::default(),
             kernel: Kernel::default(),
-            in_debug_mode,
+            in_debug_mode: false,
         }
     }
```
Comment (bobbinth) [inline assembly/src/lib.rs:98]: I like this new constructor approach! A couple of suggestions - but they are pretty minor - so, feel free to keep it as is:

- I wonder if we should get rid of `with_debug_mode()` and keep the signature of `new()` as it was before (i.e., `fn new(in_debug_mode: bool) -> Self`. We'd still keep `with_kernel()` and `with_module_provider()` as you have now.
- very minor nit: I'd move `with_kernel()` after `with_module_provider()`.

**Review by bobbinth — CHANGES_REQUESTED** [review]
I like it! Thank you! I left just two small comments inline.

File: assembly/src/lib.rs (hunk 6 above)
Comment (vlopes11) [reply assembly/src/lib.rs]: Agreed. Done!

File: assembly/src/lib.rs (hunk 7 above)
Comment (vlopes11) [reply assembly/src/lib.rs:98]: Regarding the order, I always tend to keep alphabetical order so its easier to seek, but in this case, logically, it makes more sense to have the module provider before the kernel. Done!

About the second point, we have the [builder pattern docs](https://doc.rust-lang.org/1.0.0/style/ownership/builders.html) that says we should use the `new` only with the minimal viable arguments to create the type. However, since this is an hybrid (i.e. we don't have a distinct type for the builder), then we don't necessarily have to follow these rules. wdyt?

Comment (bobbinth) [reply assembly/src/lib.rs:98]: Regarding the second point: I don't have a strong preference - so, let's keep it as you have now.

**Review by bobbinth — APPROVED** [review]
All looks good! Thank you!
